scylla/src/authentication/mod.rs

```diff
@@ -0,0 +1,96 @@
+use async_trait::async_trait;
+use bytes::{BufMut, BytesMut};
+
+/// Type to represent an authentication error message.
+pub type AuthError = String;
+
+/// Trait used to represent a user-defined custom authentication.
+#[async_trait]
+pub trait AuthenticatorSession: Send + Sync {
+    /// To handle an authentication challenge initiated by the server.
+    /// The information contained in the token parameter is authentication protocol specific.
+    /// It may be NULL or empty.
+    async fn evaluate_challenge(
+        &mut self,
+        token: Option<&[u8]>,
+    ) -> Result<Option<Vec<u8>>, AuthError>;
+
+    /// To handle the success phase of exchange.
+    /// The token parameters contain information that may be used to finalize the request.
+    async fn success(&mut self, token: Option<&[u8]>) -> Result<(), AuthError>;
+}
+
+/// Trait used to represent a factory of [`AuthenticatorSession`] instances.
+/// A new [`AuthenticatorSession`] instance will be created for each session.
+///
+/// The custom authenticator can be set using SessionBuilder::authenticator_provider method.  
+///
+/// Default: [`PlainTextAuthenticator`] is the default authenticator which requires username and
+/// password. It can be set by using SessionBuilder::user(\"user\", \"pass\") method.
+#[async_trait]
+pub trait AuthenticatorProvider: Sync + Send {
+    /// A pair of initial response and boxed [`AuthenticatorSession`]
+    /// should be returned if authentication is required by the server.
+    async fn start_authentication_session(
+        &self,
+        name: &str,
```

Comment:
The name `name` is not too descriptive, it's not obvious what the `name` refers to. How about `authenticator_name`?